ext/workspaces: port from wlr/workspaces to ext-workspace-v1 for a.o. labwc support

jp7677 opened the pull request:
This PR allows to use `wlr/workspaces` with `labwc` >= 0.8.3. This is more or less a follow up to https://github.com/Alexays/Waybar/pull/3481 and based on https://github.com/labwc/labwc/pull/2365

Additional to labwc, this protocol is also finding its way into other compositors
- niri with https://github.com/YaLTeR/niri/pull/1800
- Maomaowm with https://github.com/DreamMaoMao/maomaowm/tree/ext-workspace
- Hyprland with https://github.com/hyprwm/Hyprland/pull/10818
- Jay (https://github.com/mahkoh/jay/blob/master/docs/features.md#protocol-support)

~The heavy lifting is done in this PR, it works with labwc, but it doesn't take protocol differences between the two version into account. The last commit is a hack because workspace creation is too late and doesn't get the initial state (same reason why not the name, but the workspace id appears).~ This is now a complete reimplementation and has seen testing from multiple configurations and compositors, see discussions below.
`wayland-protocols` dependency needs to be `>=1.39` for this module to be built.

~The big outstanding question is the module name and location. The new upstream workspaces protocol is no longer Wlroots specific, so `wlr/workspaces` is probably not the best choice.~ Also moved/renamed the module from wlr/workspaces to ext/workspaces.

Obligatory screenshot:

![Screenshot](https://github.com/user-attachments/assets/cf1f99e9-b6ed-42a6-8213-6c3e64fda258)


Closes https://github.com/Alexays/Waybar/issues/3967

~Note that for CI to fully succeed, https://github.com/Alexays/Waybar/pull/4094 has to land first.~ ~The first commit is taken from https://github.com/Alexays/Waybar/pull/4094 to let CI pass and also to make my life easier by not having to revert the linting changes after running `clang-format` locally..~  Linting has been addressed om master, thanks!

jp7677 commented:
I reworked the code to move the workspace creation to the manager to be in line with the protocol. As a result the initial state hack is gone and state/name are properly picked up from labwc. I haven't tested with a multi monitor setup, but the current version works nicely for me i.c.w. labwc. That  said, I have a hard time figuring out how features like persistent-workspaces or all-ouputs are supposed to work (both functionality and code-wise). I'm also not sure if those features still makes sense with the protocol update. May be those features should be dropped? Or may be it makes more sense to start fresh, eventually as a new module? 
Leaving this for now for some feedback about the general direction.

jp7677 commented:
I'm happy and all things that needs to be done in my opinion are done. Ready for review!

jp7677 commented:
Thanks for firing up CI. The meson magic should be good now, also linting for the files this PR touches. Let's if CI agrees ;) 

If you want, I can also add a few separate commits for fixing the formatting in other modules so that CI will become completely green.

eeyrjmr commented:
very nice!!
its builds on gentoo against 0.12.0 but one tiny issue and its to do with the unicode symbols for workspaces

entry in the config:
![2025-04-27_11-04-1745749493](https://github.com/user-attachments/assets/888c0757-4b3f-44dc-afde-37379acd985c)


running wayland:
![2025-04-27_11-04-1745749527](https://github.com/user-attachments/assets/3f3ecb47-5ac9-4218-8e2a-e51d47c50472)

it using the "default" icon throughout

jp7677 commented:
> very nice!! its builds on gentoo against 0.12.0 but one tiny issue and its to do with the unicode symbols for workspaces

You need to use the names of the workspaces in the configuration section, not the id. With a default labwc setup this looks like:

```json
        "format-icons": {
                "Workspace 1": "",
                "Workspace 2": "",
                "Workspace 3": "",
                "Workspace 4": "",
                "Workspace 5": "",
                "active": "",
                "default": ""
        },
``` 

I've changed the documentation section in this PR accordingly, so thanks for pointing this out.

eeyrjmr commented:
that works! thankyou!!

earies commented:
First - thx for this @jp7677! - tested out this patch set today and see 2 noticeable issues atm

1. When using multiple outputs/monitors (extend, not clone), no matter if `sort-by-id` is set (with all other sort options = false) and `all-outputs = true`, the `sort-by-id` appears to only take effect on a single output while the other(s) retain `sort-by-name`
2. When changing output configurations (e.g. kanshi, wlr-randr) and what appeared to be resume from swayidle while waybar is using `wlr/workspaces`, waybar will segfault (few backtraces below)

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000000000055c1da in sigc::internal::signal_emit0<void, sigc::nil>::emit (impl=0x3100000000) at /usr/include/sigc++-2.0/sigc++/signal.h:786
786	     if (!impl || impl->slots_.empty()) return;
[Current thread is 1 (Thread 0x7fd2a957f980 (LWP 19200))]
(gdb) bt
#0  0x000000000055c1da in sigc::internal::signal_emit0<void, sigc::nil>::emit (impl=0x3100000000) at /usr/include/sigc++-2.0/sigc++/signal.h:786
#1  0x00007fd2ab7d1bd9 in Glib::DispatchNotifier::pipe_io_handler(Glib::IOCondition) () at /usr/lib64/libglibmm-2.4.so.1
#2  0x00007fd2ab7d3fec in Glib::IOSource::dispatch(sigc::slot_base*) () at /usr/lib64/libglibmm-2.4.so.1
#3  0x00007fd2ab7d3ba7 in Glib::Source::dispatch_vfunc(_GSource*, int (*)(void*), void*) () at /usr/lib64/libglibmm-2.4.so.1
#4  0x00007fd2ab60e7c5 in ??? () at /usr/lib64/libglib-2.0.so.0
#5  0x00007fd2ab6107b7 in ??? () at /usr/lib64/libglib-2.0.so.0
#6  0x00007fd2ab610ebc in g_main_context_iteration () at /usr/lib64/libglib-2.0.so.0
#7  0x00007fd2ab8ed78d in g_application_run () at /usr/lib64/libgio-2.0.so.0
#8  0x000000000047eeb2 in waybar::Client::main (this=this@entry=0x15b52a10, argc=<optimized out>, argc@entry=1, argv=<optimized out>, argv@entry=0x7ffdbc85eb38) at /usr/include/glibmm-2.4/glibmm/refptr.h:259
#9  0x0000000000472464 in main (argc=1, argv=0x7ffdbc85eb38) at ../src/main.cpp:107
```

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fb04fa8ea05 in Glib::DispatchNotifier::send_notification(Glib::Dispatcher*) () from /usr/lib64/libglibmm-2.4.so.1
[Current thread is 1 (Thread 0x7fb04c56e980 (LWP 19959))]
(gdb) bt
#0  0x00007fb04fa8ea05 in Glib::DispatchNotifier::send_notification(Glib::Dispatcher*) () at /usr/lib64/libglibmm-2.4.so.1
#1  0x00000000004ff689 in waybar::modules::wlr::WorkspaceManager::handle_done (this=<optimized out>) at ../src/modules/wlr/workspace_manager.cpp:120
#2  0x0000000000505bc2 in waybar::modules::wlr::workspace_manager_handle_done (data=<optimized out>, _=<optimized out>) at ../src/modules/wlr/workspace_manager_binding.cpp:45
#3  0x00007fb04e1260ca in ??? () at /usr/lib64/libffi.so.8
#4  0x00007fb04e125600 in ??? () at /usr/lib64/libffi.so.8
#5  0x00007fb04e125cbd in ffi_call () at /usr/lib64/libffi.so.8
#6  0x00007fb04fc61851 in ??? () at /usr/lib64/libwayland-client.so.0
#7  0x00007fb04fc5dbbf in ??? () at /usr/lib64/libwayland-client.so.0
#8  0x00007fb04fc5eddb in wl_display_dispatch_queue_pending () at /usr/lib64/libwayland-client.so.0
#9  0x00007fb04e9676c4 in ??? () at /usr/lib64/libgdk-3.so.0
#10 0x00007fb04e92f050 in gdk_display_get_event () at /usr/lib64/libgdk-3.so.0
#11 0x00007fb04e9673e2 in ??? () at /usr/lib64/libgdk-3.so.0
#12 0x00007fb04e5917c5 in ??? () at /usr/lib64/libglib-2.0.so.0
#13 0x00007fb04e5937b7 in ??? () at /usr/lib64/libglib-2.0.so.0
#14 0x00007fb04e593ebc in g_main_context_iteration () at /usr/lib64/libglib-2.0.so.0
#15 0x00007fb04e7df78d in g_application_run () at /usr/lib64/libgio-2.0.so.0
#16 0x000000000047eeb2 in waybar::Client::main (this=this@entry=0x105bda10, argc=<optimized out>, argc@entry=1, argv=<optimized out>, argv@entry=0x7ffdc9b0f8c8) at /usr/include/glibmm-2.4/glibmm/refptr.h:259
#17 0x0000000000472464 in main (argc=1, argv=0x7ffdc9b0f8c8) at ../src/main.cpp:107
```

jp7677 commented:
> When changing output configurations (e.g. kanshi, wlr-randr) and what appeared to be resume from swayidle while waybar is using `wlr/workspaces`, waybar will segfault (few backtraces below)

Thanks a lot for raising this. Seems that I had a small misunderstanding how the Waybar `WorkspaceManager` is supposed to work. I could reproduce this with unplugging/plugging a second monitor. I've reworked deconstruction and it should be good now. Could you please retest and report back?

>  When using multiple outputs/monitors (extend, not clone), no matter if `sort-by-id` is set (with all other sort options = false) and `all-outputs = true`, the `sort-by-id` appears to only take effect on a single output while the other(s) retain `sort-by-name`

Am I assuming correctly that you are using `labwc`? Could you please share the `desktop` part of your `labwc` desktop configuration and the `wlr/workspaces` part of your `waybar` config? The sorting works fine here, lets see if I can reproduce with your configuration.

Note though that `all-outputs` doesn't matter for `labwc` since workspaces in  `labwc` are global and not per output, so each output already gets all workspace assigned to. Also note that `labwc` doesn't send IDs for workspaces, only names, as a fallback an internal Waybar counter is used. So setting workspace names in the `labwc` configuration and using `sort-by-name` is the clearest approach. That said, I've also reworked the ID/name fallback in Waybar, those should be identical now for each output (since the counter is no longer global but per `WorkspaceManager`). Please let me know if that improves what you are seeing.

earies commented:
> I've reworked deconstruction and it should be good now. Could you please retest and report back?

Much better now - so far so good testing a few hotplug situations w/ multiple monitors

> Am I assuming correctly that you are using labwc?

Yes, sorry - testing performed against commit https://github.com/labwc/labwc/commit/cb58156904ec3e3401c8600b784485c36b978716

> Could you please share the desktop part of your labwc desktop configuration

```
  <desktops>
    <popupTime>300</popupTime>
    <number>12</number>
    <names>
      <name>1</name>
      <name>2</name>
      <name>3</name>
      <name>4</name>
      <name>5</name>
      <name>6</name>
      <name>7</name>
      <name>8</name>
      <name>9</name>
      <name>10</name>
      <name>11</name>
      <name>12</name>
    </names>
  </desktops>
```

As you can see above, the workspace "names" are integers but the prior ordering issue I was encountering was string sorting on one output (e.g. `1, 10, 11, 12, 2, 3, 4, ....`)

Now, I observe both outputs with the same ordering.  Commenting out all "sort" options gives the desired behavior but flipping any of below to `true` will result in the described above behavior

```
    "wlr/workspaces": {
        "format": "   {name}   ",
        "on-click": "activate",
        "format-icons": {
            "active": "🔳",
            "default": "⬜",
            "urgent": "🟥"
        }
        //"all-outputs": true,
        //"sort-by-coordinates": false,
        //"sort-by-name": false,
        //"sort-by-id": true
    },
```

![20250511-094337](https://github.com/user-attachments/assets/b5cb1e69-aaf8-4c4b-ac47-bdde9ba65c1d)

jp7677 commented:
> Much better now - so far so good testing a few hotplug situations w/ multiple monitors

Nice!

> Now, I observe both outputs with the same ordering. Commenting out all "sort" options gives the desired behavior but flipping any of below to `true` will result in the described above behavior

Ah right, the workspace ID and name are both strings, thus string sorting and not number sorting is applied. Only when no sorting is set, it falls back to sort by an internal ID which is an integer (https://github.com/jp7677/Waybar/blob/ext-workspaces/src/modules/wlr/workspace_manager.cpp#L158 ).  
What is wise? I can stick to the old implementation and change `sort-by-id` to actually do `sort-by-name-numerically` which would work nicely since labwc doesn't send ID's. That said, since the protocol defines ID and name, having `sort-by-id` not do sort by ID can also be confusing.

earies commented:
> PPS: I've pushed a commit (can be squashed later) that detects if numeric sorting should be applied for both name and workspace ID. In your case all sort settings except coordinates should yield numeric sorting.

Thx - latest commits LGTM (note: `"sort-by-coordinates": true` w/ all others set to false still appears to preserve numeric sorting of workspace names)

jp7677 commented:
> > PPS: I've pushed a commit (can be squashed later) that detects if numeric sorting should be applied for both name and workspace ID. In your case all sort settings except coordinates should yield numeric sorting.
> 
> Thx - latest commits LGTM (note: `"sort-by-coordinates": true` w/ all others set to false still appears to preserve numeric sorting of workspace names)

Nice. I did a last optimization, sort only when needed, but this shouldn't change anything. W.r.t. coordinates, `labwc`  doesn't send coordinates, so in that case ordering falls back again to internal ID,, which is order by creating which matches numeric sorting in your case.

jp7677 commented:
@Alexays Any chance that you could fire up CI again?

jp7677 commented:
@Alexays Gentle ping.

jp7677 commented:
> Just a small nitpick: using `wlr/workspaces` feels wrong, its the official wayland protocol for workspaces rather than a wlroots one.

Yes, I agree on that. I had hoped to get a hint from the maintainers here for a better location/name, but so far no feedback :( I'm hesitant to decide for myself how this PR fits best into this project.

YaLTeR commented:
With all-outputs false, I see workspaces only on one of the bars (for that output). Across compositor restarts I've seen it use either. 

With all-outputs true, I see the same set of workspaces (for that one output) on both.

jp7677 commented:
Could you please post the waybar logs prefixed with `[wlr/workspaces]` when running waybar with debug logging (`waybar -l debug`)?

YaLTeR commented:
Seems to work well!

YaLTeR commented:
Btw I also think that renaming the module to `ext/workspace` would make sense. Especially since `wlr/workspace` hasn't been enabled by default yet I think?

jp7677 commented:
> Btw I also think that renaming the module to `ext/workspace` would make sense. Especially since `wlr/workspace` hasn't been enabled by default yet I think?

Yes I agree. As stated above, I’m looking for a hint from the maintainers here how to handle this (naming, file locations, etc) in the best interest of the project.

Consolatis commented:
I'd just `s/wlr/ext/g` and call it a day.

jp7677 commented:
> I'd just `s/wlr/ext/g` and call it a day.

Alright, I just did exactly that.

jp7677 commented:
> When marking the workspace as hidden, is it allowed to completely hide the workspace button on waybar?

@DreamMaoMao I've added a commit which hides hidden workspaces by default, but also added an option `ignore-hidden` (default `true`) which can be set to `false` to still show hidden workspaces. Does this works for you?

DreamMaoMao commented:
> > When marking the workspace as hidden, is it allowed to completely hide the workspace button on waybar?
> 
> @DreamMaoMao I've added a commit which hides hidden workspaces by default, but also added an option `ignore-hidden` (default `true`) which can be set to `false` to still show hidden workspaces. Does this works for you?

look good to me

DreamMaoMao commented:
@jp7677 

![image](https://github.com/user-attachments/assets/d56c9aa2-d885-470c-8536-64a930564f0f)

i have a question. why these two global variables are added. It seems to be just for printing a number on the debug log. But if I temporarily destroy and recreate some workspaces, it simply incrementing. The meaning it represents doesn't seem correct.

![image](https://github.com/user-attachments/assets/916c2970-01dd-40af-afbc-df8c999e1693)

jp7677 commented:
@Alexays Gentle ping. Is there anything I can do to get you enthusiastic about this PR?

jp7677 commented:
Hi @Alexays I guess you have your valid reasons to ignore this PR, but please be aware that giving no feedback at all is imho the worst feedback of all. I could completely understand if you don’t see this PR fit into your project or that you don’t want to review a large PR like this one. Whatever the reason, I would very much preferred a firm “no”, preferably in the early stages, instead of no feedback at all. Due to the latter I had the idea that pushing forward is worth it and I can get this into a mergable state. Well, it seems I have failed. My idea is now that all those hours of research and coding for a feature I wanted to have, but also for filling a small gap in the Wayland ecosystem, were for nothing. This doesn’t feel very good :(. I’m also sorry for others who put time and energy into testing or giving feedback. Anyway, thanks for reading.

PS: your readme says “Contributions welcome”, may be wise to update that part for managing expectations, it seems that I misunderstood that part.

jp7677 commented:
Thanks, I’ve edited the last part, indeed wise to not assume anything.

sfzaw commented:
@jp7677 i understand the frustration you are going through as i am going through it at the same time. dont lose patience and just know we are all here. your contribution is definitely appreciated. just as alanon says, without ur pr i wouldve switched to sfwbar too. (in the meantime, see if you can improve it some other as you did before while waiting... may be take on other compositors ... just throwing an idea...)

jp7677 commented:
Thanks a lot for the merge!
